libwild/src/args.rs

```diff
@@ -1090,6 +714,1414 @@ fn warn_unsupported(opt: &str) -> Result {
+        }
+    }
+}
+
+impl<'a> PrefixOptionDeclaration<'a> {
+    #[must_use]
+    pub fn help(mut self, text: &'static str) -> Self {
+        self.help_text = text;
+        self
+    }
+
+    #[must_use]
+    pub fn sub_option(mut self, name: &'static str, help: &'static str) -> Self {
+        self.sub_options.insert(name, help);
+        self
+    }
+
+    pub fn execute(self, handler: fn(&mut Args, &mut Vec<Modifiers>, &str) -> Result<()>) {
+        let prefix_handler = PrefixOptionHandler {
+            help_text: self.help_text,
+            sub_options: self.sub_options,
+            handler,
+        };
+
+        self.parser
+            .prefix_options
+            .insert(self.prefix, prefix_handler);
+    }
+}
+
+fn strip_option(arg: &str) -> Option<&str> {
+    arg.strip_prefix("--").or(arg.strip_prefix('-'))
+}
+
+fn setup_argument_parser() -> ArgumentParser {
+    let mut parser = ArgumentParser::new();
+
+    parser
+        .declare_prefix("L")
+        .help("Add directory to library search path")
+        .execute(|args, _modifier_stack, value| {
+            let handle_sysroot = |path| {
+                args.sysroot
+                    .as_ref()
+                    .and_then(|sysroot| maybe_forced_sysroot(path, sysroot))
+                    .unwrap_or_else(|| Box::from(path))
+            };
+
+            let dir = handle_sysroot(Path::new(value));
+            args.save_dir.handle_file(value)?;
+            args.lib_search_path.push(dir);
+            Ok(())
+        });
+
+    parser
+        .declare_prefix("l")
+        .help("Link with library")
+        .sub_option(":filename", "Link with specific file")
+        .sub_option("libname", "Link with library libname.so or libname.a")
+        .execute(|args, modifier_stack, value| {
+            let spec = if let Some(stripped) = value.strip_prefix(':') {
+                InputSpec::File(Box::from(Path::new(stripped)))
+            } else {
+                InputSpec::Lib(Box::from(value))
+            };
+            args.inputs.push(Input {
+                spec,
+                search_first: None,
+                modifiers: *modifier_stack.last().unwrap(),
+            });
+            Ok(())
+        });
+
+    parser
+        .declare_prefix("u")
+        .help("Force undefined symbol")
+        .execute(|args, _modifier_stack, value| {
+            args.undefined.push(value.to_owned());
+            Ok(())
+        });
+
+    parser
+        .declare_prefix("m")
+        .help("Set target architecture")
+        .sub_option("elf_x86_64", "x86-64 ELF target")
+        .sub_option("aarch64elf", "AArch64 ELF target")
+        .sub_option("elf64lriscv", "RISC-V 64-bit ELF target")
+        .execute(|args, _modifier_stack, value| {
+            args.arch = Architecture::from_str(value)?;
+            Ok(())
+        });
+
+    parser
+        .declare_prefix("z")
+        .help("Linker option")
+        .sub_option("now", "Set BIND_NOW for object")
```

Comment:
It could be good if `sub_option` took an execute callback as an argument. That way we can be sure that the help and actually supported options won't get out-of-sync.